db/commitlog/commitlog.cc

```diff
@@ -3135,6 +3555,100 @@ db::commitlog::read_log_file(sstring filename, sstring pfx, commit_load_reader_f
+                crc.process(id);
+                crc.process(off);
+                crc.process(rem);
+
+                if (crc.checksum() != checksum) {
+                    auto slack = next - pos;
+                    if (size != 0) {
+                        clogger.debug("Fractured segment entry at {} has broken header. Skipping to next chunk ({} bytes)", rp, slack);
+                        corrupt_size += slack;
+                    }
+                    co_await skip_to_chunk(next);
+                    co_return;
+                }
+
+                buf = co_await read_data(actual_size - entry_header_size - segment::fragmented_entry_overhead_size);
+
+                using entry_fragment = db::commitlog::replay_state::impl::entry_fragment;
+                entry_fragment frag;
+                frag.offset = off;
+                frag.rem = rem;
+                frag.end = off + buf.size_bytes();
+                frag.rpbuf = buffer_and_replay_position{std::move(buf), rp};
+
+                clogger.debug("fragment id={} off={}, end={}, rem={} ", id, off, frag.end, rem);
+
+                auto& frag_states = state.fragment_state[id];
+                bool done = false;
+
+                auto join = [](entry_fragment& f1, entry_fragment& f2) {
+                    auto size1 = f1.rpbuf.buffer.size_bytes();
+                    auto size2 = f2.rpbuf.buffer.size_bytes();
+                    auto data1 = std::move(f1.rpbuf.buffer).release();
+                    auto data2 = std::move(f2.rpbuf.buffer).release();
+                    for (auto&& bb : data2) {
+                        data1.emplace_back(std::move(bb));
+                    }
+                    f1.rem = f2.rem;
+                    f1.end = f2.end;
+                    f1.rpbuf.buffer = fragmented_temporary_buffer(std::move(data1), size1+size2);
+                };
+
+                for (auto& fs : frag_states) {
+                    if (fs.end == frag.offset) {
+                        join(fs, frag);
+                        done = true;
+                    } else if (frag.end == fs.offset) {
+                        join(frag, fs);
+                        fs = std::move(frag);
+                        done = true;
+                    }
+                    if (done) {
+                        break;
+                    }
+                }
+                if (!done) {
+                    frag_states.emplace_back(std::move(frag));
+                }
+
+                std::sort(frag_states.begin(), frag_states.end(), [](auto& s1, auto& s2) { return s1.offset < s2.offset; });
+                while (frag_states.size() > 1) {
+                    auto& f1 = frag_states[frag_states.size() - 2];
+                    auto& f2 = frag_states.back();
+                    if (f1.end == f2.offset) {
+                        join(f1, f2);
+                        frag_states.pop_back();
+                        continue;
+                    }
+                    break;
```

Comment:
This is a hard to read loop, since the normal case is to break, and the special case is to loop.

I'm sure it can be simplified by joining when we have all the fragments ready.

Reply:
Simplified + made perhaps more logical for less continue-obsessed (a weakness, I know...) people. Dropped the code trying to avoid adding to the vector at all in favour of just sort + reduce at end loop.